include/boost/iostreams/filter/zlib.hpp

```diff
@@ -347,7 +347,7 @@ bool zlib_compressor_impl<Alloc>::filter
     int result = xdeflate(flush ? zlib::finish : zlib::no_flush);
     after(src_begin, dest_begin, true);
     zlib_error::check BOOST_PREVENT_MACRO_SUBSTITUTION(result);
-    return result != zlib::stream_end; 
+    return result == zlib::okay; 
```

Comment:
_Since we return false in that case, rather than throwing an exception as previously, doesn't that mean that iostreams treats the operations as completing successfully where it really needs to be called again._
No it does not. Because the filter function does not guarantee to complete the job in a single call, instead the outer functions check the end of stream and continue to call this function if it is not reached. Also user should know that the read and write functions does not guarantee to complete the job in a single call due to buffer issues. User should always check the completeness. So that it's meaningless to throw an exception for a situation already have to be known.